### PR TITLE
Cleaned up compile script, made more debuggable 

### PR DESCRIPTION
**Pull request by Mainstay-Noah-Huppert**
I went hunting around the download server for get-pip.py and found a URL which is pinned by version, I think. It hasn't been updated since Feb. so I think it won't change.

### CODE REVIEW

File: bin/compile [hunk 1]
```diff
@@ -40,18 +40,19 @@ puts_step 'install pip'
...
-readonly install_pip_script_sha256="fa6f3fb93cce234cd4e8dd2beb54a51ab9c247653b52855a48dd44e6b21ff28b"
+readonly install_pip_script_url="https://bootstrap.pypa.io/pip/3.5/get-pip.py"
+readonly install_pip_script_sha256="311afebb7cdd310eb3a3a6bb6fffef53d84493db98c7cebf4008a18d3418c8be"
 
 readonly install_pip_dir="$HOME/.local/bin"
 mkdir -p "$install_pip_dir"
 
-curl -L --create-dirs "https://bootstrap.pypa.io/get-pip.py" -o "$install_pip_script_file"
+curl -L --create-dirs "$install_pip_script_url" -o "$install_pip_script_file"
 echo "$install_pip_script_sha256 $install_pip_script_file" | sha256sum --check
```
Comment (edanaher) [inline bin/compile:50]: Can we get a more precise error message to make it easier to track down future issues?  Maybe something like:

```suggestion
if ! echo "$install_pip_script_sha256 $install_pip_script_file" | sha256sum --check; then
 echo "Pip install script doesn't match expected sha; update buildpack-marshall/bin/compile appropriately"
 exit 4
fi
```

Comment (Mainstay-Noah-Huppert) [reply bin/compile:50]: Yea sure, while I'm here I'll give some of the other areas of the script similar treatment.

File: bin/common.sh [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+# Download and source the Heroku buildpack bash helper library
+load_heroku_buildpack_stdlib() {
+    if [[ ! -f "$HEROKU_BUILDPACK_STDLIB_FILE" ]]; then
+	   download_and_checksum "$HEROKU_BUILDPACK_STDLIB_URL" "$HEROKU_BUILDPACK_STDLIB_FILE" "$HEROKU_BUILDPACK_STDLIB_SHA256SUM"
+    fi
+
+    # Tell shellcheck not to try and lint this file which we will download
+    # shellcheck disable=SC1090
+    source "$HEROKU_BUILDPACK_STDLIB_FILE"
+}
```
Comment (Mainstay-Noah-Huppert) [inline bin/common.sh:164]: This file has been code reviewed previously when it was merged into the Marshall repository. With the exception of these new helper functions.

**Comment by Mainstay-Noah-Huppert** [comment]
Greatly expanded the scope of this PR. But I think the buildpack script is in a much better place now. Every step that can fail has a unique exit code and error message. I also added a GitHub workflow to run a tool named ShellCheck, which statically analyses the bash scripts and catches errors.

I've deployed a review app using this PR's branch, and it worked: https://dashboard.heroku.com/apps/marshall-review-pr-5925/activity/builds/f06c5bb0-c8b5-44c1-99bf-bbd7c0d79687

**Review by edanaher — COMMENTED** [review]
I haven't gotten through bin/compile yet, but wanted to give you a first round of feedback rather than waiting to finish that.

Overall I generally like this; it's just a lot of bash.

File: bin/compile [hunk 3]
```diff
@@ -2,156 +2,240 @@
...
-curl --silent --retry 5 --retry-max-time 15 -L "https://github.com/yarnpkg/yarn/releases/download/v1.15.0/yarn-v1.15.0.tar.gz" -o /tmp/yarn.tar.gz
-tar zxf /tmp/yarn.tar.gz -C .heroku/yarn --strip 1
-
-echo 'yarn version:' $(yarn --version)
-echo 'node version:' $(node --version)
+# add to path
+export PATH="$PATH:$NODEJS_INSTALL_DIR/bin:$YARN_INSTALL_DIR/bin"
+
+# NodeJS
+run_check "mkdir -p $NODEJS_INSTALL_DIR" \
```
Comment (DemiAHMS) [inline bin/compile:144]: Here and everywhere else, consider making the error message also be a read-only string defined at the top of your scripts

**Review by DemiAHMS — APPROVED** [review]

**Comment by Mainstay-Noah-Huppert** [comment]
> I haven't gotten through bin/compile yet, but wanted to give you a first round of feedback rather than waiting to finish that.
> 
> Overall I generally like this; it's just a lot of bash.

Thanks, yea sorry it's a lot 😅

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,6 @@
+.PHONY: check
```
[the remark this answers is not in the document]
Comment (Mainstay-Noah-Huppert) [reply Makefile:1]: Yep sounds good

File: bin/common.sh [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        die "$EXIT_REQUIRED_PROGRAMS_MISSING" "Required program(s) missing: ${missing_programs[*]}"
+    fi
+}
+
+# Returns TRUE base_path is newer (not inclusive) than compare_path, FALSE otherwise.
+path_newer() { # ( base_path, compare_path )
+    local -r base_path="$1"
+    local -r compare_path="$2"
+
+    local -ri base_age=$(run_check "date -r $base_path +%s" "$EXIT_PATH_NEWER_DATE_FAIL" "Failed to get age of '$base_path' path")
```
[the remark this answers is not in the document]
Comment (Mainstay-Noah-Huppert) [reply bin/common.sh:89]: I believe BSD date can use `-r`.

```shell
$ echo "I am using BSD date in this shell"
$ man date | grep BSD
DATE(1)                   BSD General Commands Manual                  DATE(1)
BSD                               May 7, 2015                              BSD
$ echo "Some file" > foo
$ % date -r foo '+%s'
1633102606
```

**Comment by Mainstay-Noah-Huppert** [comment]
Thanks for for the reviews. I will address them next week 👍

File: bin/common.sh [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    local -ra required_programs=("$@")
+
+    local -a missing_programs=()
+    for required_program in "${required_programs[@]}"; do
+        if ! which "$required_program" &> /dev/null; then
+            missing_programs+=("$required_program")
+        fi
+    done
+
+    if [[ -n "${missing_programs[*]}" ]]; then
```
[the remark this answers is not in the document]
Comment (Mainstay-Noah-Huppert) [reply bin/common.sh:79]: Will change, I think I wrote this function before I got in the habit of checking the size of arrays in bash.

**Review by DemiAHMS — DISMISSED** [review]

**Review by chdsbd — APPROVED** [review]

**Pull request merged**